src/test/kotlin/uk/gov/dluhc/emsintegrationapi/service/RetrieveGssCodeServiceTest.kt

```diff
@@ -0,0 +1,181 @@
+
+    @InjectMocks
+    private lateinit var retrieveGssCodeService: RetrieveGssCodeService
+
+    @Nested
+    inner class GetGssCodeFromCertificateSerial {
+
+        @Test
+        fun `should throw IER not found exception given IER API client throws IER not found exception`() {
+            // Given
+            val certificateSerial = "123456789"
+
+            val expected = IerEroNotFoundException(certificateSerial)
+            given(ierApiClient.getEroIdentifier(any())).willThrow(expected)
+
+            // When
+            val ex = catchThrowableOfType(
+                { retrieveGssCodeService.getGssCodeFromCertificateSerial(certificateSerial) },
+                IerEroNotFoundException::class.java
+            )
+
+            // Then
+            assertThat(ex).isEqualTo(expected)
+            verify(ierApiClient).getEroIdentifier(certificateSerial)
+            verifyNoInteractions(eroService)
+        }
+
+        @Test
+        fun `should throw general IER exception given IER API client throws general exception`() {
+            // Given
+            val certificateSerial = "123456789"
+
+            val expected = IerGeneralException("Error getting eroId for certificate serial $certificateSerial")
+            given(ierApiClient.getEroIdentifier(any())).willThrow(expected)
+
+            // When
+            val ex = catchThrowableOfType(
+                { retrieveGssCodeService.getGssCodeFromCertificateSerial(certificateSerial) },
+                IerGeneralException::class.java
+            )
+
+            // Then
+            assertThat(ex).isEqualTo(expected)
+            verify(ierApiClient).getEroIdentifier(certificateSerial)
+            verifyNoInteractions(eroService)
+        }
+
+        @Test
+        fun `should throw ERO not found exception given ERO API client throws not found exception`() {
+            // Given
+            val certificateSerial = "123456789"
+            val eroIdFromIerApi = getRandomEroId()
+            val expected = ElectoralRegistrationOfficeNotFoundException(certificateSerial)
+
+            given(ierApiClient.getEroIdentifier(any())).willReturn(EROCertificateMapping(eroId = eroIdFromIerApi, certificateSerial = certificateSerial))
+            given(eroService.lookupGssCodesForEro(any())).willThrow(expected)
+
+            // When
+            val ex = catchThrowableOfType(
+                { retrieveGssCodeService.getGssCodeFromCertificateSerial(certificateSerial) },
+                ElectoralRegistrationOfficeNotFoundException::class.java
+            )
+
+            // Then
+            assertThat(ex).isEqualTo(expected)
+            verify(ierApiClient).getEroIdentifier(certificateSerial)
+            verify(eroService).lookupGssCodesForEro(eroIdFromIerApi)
+        }
+
+        @Test
+        fun `should throw general ERO exception given ERO API client throws general exception`() {
+            // Given
+            val certificateSerial = "123456789"
```

Comment:
Minor, may consider having a constant.  This initiated in several places and is the same.

Reply:
Done, Thanks, I did not try to change existing code, copied as is from RegisterChecker. There are a few more improvements to be done with the code copied, will include in the later PR